src/main/java/org/commcare/formplayer/services/FormDefinitionService.java

```diff
@@ -54,4 +67,102 @@ public SerializableFormDefinition getOrCreateFormDefinition(
+     * Ensure raw xml is accessible locally in case serialization changes which would break the ability to
+     * deserialize the serialized form def in postgres
+     *
+     * @return True if the value was written to storage or False if it already exists in storage.
+     */
+    public boolean writeToLocalStorage(FormDef formDef) {
+        this.storageFactory.getStorageManager().registerStorage(FormDef.STORAGE_KEY, FormDef.class);
+        String xmlns = formDef.getMainInstance().schema;
+        if (getFormDefFromStorage(xmlns).isPresent()) {
+            return false;
+        }
+        writeFormDefToStorage(formDef);
+        return true;
+    }
+
+    /**
+     * This method uses the sessionId to cache deserialized FormDefs
+     * Deserializing a serialized FormDef object is costly, and this avoids doing so in between requests
+     * within the same session
+     *
+     * @param session session that contains session id and serialized formDef
+     * @return deserialized FormDef object
+     */
+    public FormDef getFormDef(SerializableFormSession session) {
+        FormDef formDef = this.getFormDefWithCache(session);
+        // ensure previous tree references are cleared (only necessary when retrieving from cache)
+        formDef.getMainInstance().cleanCache();
+        return formDef;
+    }
+
+    /**
+     * Always use public getFormDef to ensure internal FormDef cached references are cleared.
+     *
+     * We can't use @{@link Cacheable} here since we want to keep this method private.
+     */
+    private FormDef getFormDefWithCache(SerializableFormSession session) {
+        Cache cache = caches.getCache("form_definition");
+        return cache.get(session.getId(), () -> getFormDefFromSession(session));
+    }
+
+    private FormDef getFormDefFromSession(SerializableFormSession session) {
+        SerializableFormDefinition formDefinition = session.getFormDefinition();
+        try {
+            return FormDefStringSerializer.deserialize(formDefinition.getSerializedFormDef());
+        } catch (Exception e) {
+            String xmlns = formDefinition.getFormXmlns();
+            FormDef formDef = getFormDefFromStorage(xmlns).orElseThrow(() -> {
+                return new WrappedException("Unable to load form def after serialization error", e);
+            });
+            this.updateFormDefinition(formDefinition, formDef);
+            return formDef;
+        }
+    }
+
+    /**
+     * Cache the form def for future requests in this session
+     *
+     * @param session grab the id to cache on and the deserialized form def from the session
+     * @return deserialized FormDef object
+     */
+    @CachePut(key = "#session.getSessionId()")
+    public FormDef cacheFormDef(FormSession session) {
+        return session.getFormDef();
+    }
+
+    private Optional<FormDef> getFormDefFromStorage(String xmlns) {
+        this.storageFactory.getStorageManager().registerStorage(FormDef.STORAGE_KEY, FormDef.class);
```

Comment:
doesn't `storageFactory` already have the storage registered ?

Reply:
no, that's the whole crux of the matter. Specifically for SMS forms there is no app install and so the storage never get's registered.

Reply:
A `CommCarePlatform` is setup during initialization of a FormSession, but that does not register FormDef storage, only [registers PropertyManager storage](https://github.com/dimagi/commcare-core/blob/9a4192a98e640b1fcf0e8b1fc67b13c4f663de96/src/main/java/org/commcare/util/CommCarePlatform.java#L56). This happens _after_ we need access to a local form def, so we need to register the FormDef storage prior to this point anyway. 

In the context of the FormController, storage is configured due to the `@ConfigureStorageFromSession` decorator used on what appears to be all endpoints except sms. Registering FormDef storage after configuring the factory [here](https://github.com/dimagi/formplayer/blob/master/src/main/java/org/commcare/formplayer/aspects/ConfigureStorageFromSessionAspect.java#L29) makes the most sense to me.

I've been reproducing locally by modifying the FormDefinitionService to attempt to access the local storage on any `getFormDef`, though you really just need to be able to inspect the storage factory at that point. You'll see that the StorageManager returned from the factory does not have any registered storage yet.